{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ffad1565-9e98-4be0-939c-c96fe07e3448",
   "metadata": {},
   "source": [
    "# Cross-Species Gene Symbol Mapping\n",
    "\n",
    "The majority of prior knowledge in OmniPath {cite:p}`omnipath`\n",
    "(and similar databases) is based on human data and thus uses\n",
    "human gene symbols.\n",
    "\n",
    "However, gene homology can be used to convert these symbols to those\n",
    "of other organisms using `decoupler`.\n",
    "To achieve this, it uses orthology tables extracted from the HCOP\n",
    "database {cite:p}`hcop`.\n",
    "\n",
    "The following organisms are currently supported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a21bf1d0-746d-408c-b734-97eca464245e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['anole_lizard',\n",
       " 'c.elegans',\n",
       " 'cat',\n",
       " 'cattle',\n",
       " 'chicken',\n",
       " 'chimpanzee',\n",
       " 'dog',\n",
       " 'fruitfly',\n",
       " 'horse',\n",
       " 'macaque',\n",
       " 'mouse',\n",
       " 'opossum',\n",
       " 'pig',\n",
       " 'platypus',\n",
       " 'rat',\n",
       " 's.cerevisiae',\n",
       " 's.pombe',\n",
       " 'xenopus',\n",
       " 'zebrafish']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import decoupler as dc\n",
    "\n",
    "dc.op.show_organisms()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f8cb584-4251-4a56-a2ba-fdd529f21fb3",
   "metadata": {},
   "source": [
    "To demonstrate this functionality in `decoupler`,\n",
    "here the SIGNOR database (which uses human gene symbols) is loaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "361cc5f7-057d-4706-886d-dd1d94560bbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>Cell cycle: G2/M phase transition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACE</td>\n",
       "      <td>Focal segmental glomerulosclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTN1</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTN1</td>\n",
       "      <td>Glutamatergic synapse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2943</th>\n",
       "      <td>YY1</td>\n",
       "      <td>NOTCH Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2944</th>\n",
       "      <td>ZAP70</td>\n",
       "      <td>T cell activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2945</th>\n",
       "      <td>ZAP70</td>\n",
       "      <td>P38 Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946</th>\n",
       "      <td>ZBTB16</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2947</th>\n",
       "      <td>ZFYVE9</td>\n",
       "      <td>TGF-beta Signaling</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2948 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genesymbol                             pathway\n",
       "                                                   \n",
       "0          ABL1   Cell cycle: G2/M phase transition\n",
       "1           ACE  Focal segmental glomerulosclerosis\n",
       "2          ACTB                       Axon guidance\n",
       "3         ACTN1                       Axon guidance\n",
       "4         ACTN1               Glutamatergic synapse\n",
       "...         ...                                 ...\n",
       "2943        YY1                     NOTCH Signaling\n",
       "2944      ZAP70                   T cell activation\n",
       "2945      ZAP70                       P38 Signaling\n",
       "2946     ZBTB16              Acute Myeloid Leukemia\n",
       "2947     ZFYVE9                  TGF-beta Signaling\n",
       "\n",
       "[2948 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "net = dc.op.resource(\"SIGNOR\")\n",
    "net"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61d4374a-8776-4dee-bae3-95a9571bd2da",
   "metadata": {},
   "source": [
    "The obtained resource can easily be converted to mouse gene symbols."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dcea69a-19da-4348-9277-72d8c06d6b03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abl1</td>\n",
       "      <td>Cell cycle: G2/M phase transition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ace</td>\n",
       "      <td>Focal segmental glomerulosclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Actb</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Actg1</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Actn1</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1436</th>\n",
       "      <td>Yy1</td>\n",
       "      <td>NOTCH Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>Zap70</td>\n",
       "      <td>T cell activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>Zap70</td>\n",
       "      <td>P38 Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>Zbtb16</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>Zfyve9</td>\n",
       "      <td>TGF-beta Signaling</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1441 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genesymbol                             pathway\n",
       "0          Abl1   Cell cycle: G2/M phase transition\n",
       "1           Ace  Focal segmental glomerulosclerosis\n",
       "2          Actb                       Axon guidance\n",
       "3         Actg1                       Axon guidance\n",
       "4         Actn1                       Axon guidance\n",
       "...         ...                                 ...\n",
       "1436        Yy1                     NOTCH Signaling\n",
       "1437      Zap70                   T cell activation\n",
       "1438      Zap70                       P38 Signaling\n",
       "1439     Zbtb16              Acute Myeloid Leukemia\n",
       "1440     Zfyve9                  TGF-beta Signaling\n",
       "\n",
       "[1441 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_net = dc.op.translate(\n",
    "    net,\n",
    "    target_organism=\"mouse\",\n",
    ")\n",
    "m_net"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d06f6463-09cb-4ced-96b5-beed716d7372",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "    \n",
    "**Note**\n",
    "\n",
    "Homology conversion may result in the gain or loss\n",
    "of certain genes when mapping between organisms.\n",
    "Adjust the `one_to_many` parameter to make the behavior more or less strict.\n",
    "\n",
    "</div>\n",
    "\n",
    "Next, the conversion is performed for the fruit fly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "837849e3-24ba-4cc1-8f44-86dc670d1a68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abl</td>\n",
       "      <td>Cell cycle: G2/M phase transition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ance</td>\n",
       "      <td>Focal segmental glomerulosclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acer</td>\n",
       "      <td>Focal segmental glomerulosclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ance-2</td>\n",
       "      <td>Focal segmental glomerulosclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ance-3</td>\n",
       "      <td>Focal segmental glomerulosclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>yki</td>\n",
       "      <td>Hippo Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1025</th>\n",
       "      <td>14-3-3zeta</td>\n",
       "      <td>SAPK/JNK Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1026</th>\n",
       "      <td>pho</td>\n",
       "      <td>NOTCH Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027</th>\n",
       "      <td>phol</td>\n",
       "      <td>NOTCH Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1028</th>\n",
       "      <td>Sara</td>\n",
       "      <td>TGF-beta Signaling</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1029 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genesymbol                             pathway\n",
       "0            Abl   Cell cycle: G2/M phase transition\n",
       "1           Ance  Focal segmental glomerulosclerosis\n",
       "2           Acer  Focal segmental glomerulosclerosis\n",
       "3         Ance-2  Focal segmental glomerulosclerosis\n",
       "4         Ance-3  Focal segmental glomerulosclerosis\n",
       "...          ...                                 ...\n",
       "1024         yki                     Hippo Signaling\n",
       "1025  14-3-3zeta                  SAPK/JNK Signaling\n",
       "1026         pho                     NOTCH Signaling\n",
       "1027        phol                     NOTCH Signaling\n",
       "1028        Sara                  TGF-beta Signaling\n",
       "\n",
       "[1029 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_net = dc.op.translate(\n",
    "    net,\n",
    "    target_organism=\"fruitfly\",\n",
    ")\n",
    "f_net"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01d550a6-f134-4619-b7a7-e0d11a9b27d3",
   "metadata": {},
   "source": [
    "Additionaly, all database functions in `decoupler` directly accept the parametter organism,\n",
    "which under the hood it runs `decoupler.op.translate`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d44d4f1d-12c2-4342-bc6d-a6d714756e1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>abl1</td>\n",
       "      <td>Cell cycle: G2/M phase transition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ace</td>\n",
       "      <td>Focal segmental glomerulosclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>actb2</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>actb1</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>actn1</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>zap70</td>\n",
       "      <td>T cell activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1982</th>\n",
       "      <td>zap70</td>\n",
       "      <td>P38 Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>zbtb16a</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1984</th>\n",
       "      <td>zbtb16b</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>zfyve9a</td>\n",
       "      <td>TGF-beta Signaling</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1986 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genesymbol                             pathway\n",
       "0          abl1   Cell cycle: G2/M phase transition\n",
       "1           ace  Focal segmental glomerulosclerosis\n",
       "2         actb2                       Axon guidance\n",
       "3         actb1                       Axon guidance\n",
       "4         actn1                       Axon guidance\n",
       "...         ...                                 ...\n",
       "1981      zap70                   T cell activation\n",
       "1982      zap70                       P38 Signaling\n",
       "1983    zbtb16a              Acute Myeloid Leukemia\n",
       "1984    zbtb16b              Acute Myeloid Leukemia\n",
       "1985    zfyve9a                  TGF-beta Signaling\n",
       "\n",
       "[1986 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dc.op.resource(\"SIGNOR\", organism=\"zebrafish\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1cf8b880-896b-4bd2-a3fa-cb50b975d5c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Androgen</td>\n",
       "      <td>tmprss2</td>\n",
       "      <td>11.490631</td>\n",
       "      <td>2.384806e-47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Androgen</td>\n",
       "      <td>nkx3-1</td>\n",
       "      <td>10.622551</td>\n",
       "      <td>2.205102e-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Androgen</td>\n",
       "      <td>mboat2</td>\n",
       "      <td>10.472733</td>\n",
       "      <td>4.632376e-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Androgen</td>\n",
       "      <td>SLC38A4</td>\n",
       "      <td>7.363805</td>\n",
       "      <td>1.253071e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Androgen</td>\n",
       "      <td>mtmr9</td>\n",
       "      <td>6.130646</td>\n",
       "      <td>2.534403e-38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43409</th>\n",
       "      <td>p53</td>\n",
       "      <td>enpp2</td>\n",
       "      <td>2.771405</td>\n",
       "      <td>4.993215e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43410</th>\n",
       "      <td>p53</td>\n",
       "      <td>arrdc4</td>\n",
       "      <td>3.494328</td>\n",
       "      <td>4.996747e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43411</th>\n",
       "      <td>p53</td>\n",
       "      <td>myo1b</td>\n",
       "      <td>-1.148057</td>\n",
       "      <td>4.997905e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43412</th>\n",
       "      <td>p53</td>\n",
       "      <td>ctsc</td>\n",
       "      <td>-1.784693</td>\n",
       "      <td>4.998864e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43413</th>\n",
       "      <td>p53</td>\n",
       "      <td>naa50</td>\n",
       "      <td>-1.435013</td>\n",
       "      <td>4.998884e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43414 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         source   target     weight          padj\n",
       "0      Androgen  tmprss2  11.490631  2.384806e-47\n",
       "1      Androgen   nkx3-1  10.622551  2.205102e-44\n",
       "2      Androgen   mboat2  10.472733  4.632376e-44\n",
       "3      Androgen  SLC38A4   7.363805  1.253071e-39\n",
       "4      Androgen    mtmr9   6.130646  2.534403e-38\n",
       "...         ...      ...        ...           ...\n",
       "43409       p53    enpp2   2.771405  4.993215e-02\n",
       "43410       p53   arrdc4   3.494328  4.996747e-02\n",
       "43411       p53    myo1b  -1.148057  4.997905e-02\n",
       "43412       p53     ctsc  -1.784693  4.998864e-02\n",
       "43413       p53    naa50  -1.435013  4.998884e-02\n",
       "\n",
       "[43414 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dc.op.progeny(organism=\"anole_lizard\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "faa01195-189f-4635-aa27-147d13ac67f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "      <th>resources</th>\n",
       "      <th>references</th>\n",
       "      <th>sign_decision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPT15</td>\n",
       "      <td>TOA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ExTRI</td>\n",
       "      <td>10078202;10523649;10581267;10617594;10675336;1...</td>\n",
       "      <td>default activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TOA1</td>\n",
       "      <td>SPT15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TRRUST</td>\n",
       "      <td>12818428</td>\n",
       "      <td>default activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MOT1</td>\n",
       "      <td>SPT15</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>TRRUST</td>\n",
       "      <td>14988402;15509807;16858867;20627952</td>\n",
       "      <td>PMID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPT15</td>\n",
       "      <td>MOT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ExTRI</td>\n",
       "      <td>10082549</td>\n",
       "      <td>default activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MCM1</td>\n",
       "      <td>MCM1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ExTRI;NTNU.Curated</td>\n",
       "      <td>10330138;10602487;15531578;17629633;8663310;92...</td>\n",
       "      <td>PMID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>COY1</td>\n",
       "      <td>RDI1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Pavlidis2021</td>\n",
       "      <td>19635798</td>\n",
       "      <td>PMID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>COY1</td>\n",
       "      <td>POL1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Pavlidis2021</td>\n",
       "      <td>12438259;12665598;18347061</td>\n",
       "      <td>PMID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>HAP2</td>\n",
       "      <td>SSA4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Pavlidis2021</td>\n",
       "      <td>24041570</td>\n",
       "      <td>PMID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>HAP3</td>\n",
       "      <td>SSA4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Pavlidis2021</td>\n",
       "      <td>24041570</td>\n",
       "      <td>PMID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>HAP5</td>\n",
       "      <td>SSA4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Pavlidis2021</td>\n",
       "      <td>24041570</td>\n",
       "      <td>PMID</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>367 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    source target  weight           resources  \\\n",
       "0    SPT15   TOA1     1.0               ExTRI   \n",
       "1     TOA1  SPT15     1.0              TRRUST   \n",
       "2     MOT1  SPT15    -1.0              TRRUST   \n",
       "3    SPT15   MOT1     1.0               ExTRI   \n",
       "4     MCM1   MCM1     1.0  ExTRI;NTNU.Curated   \n",
       "..     ...    ...     ...                 ...   \n",
       "362   COY1   RDI1     1.0        Pavlidis2021   \n",
       "363   COY1   POL1     1.0        Pavlidis2021   \n",
       "364   HAP2   SSA4     1.0        Pavlidis2021   \n",
       "365   HAP3   SSA4     1.0        Pavlidis2021   \n",
       "366   HAP5   SSA4     1.0        Pavlidis2021   \n",
       "\n",
       "                                            references       sign_decision  \n",
       "0    10078202;10523649;10581267;10617594;10675336;1...  default activation  \n",
       "1                                             12818428  default activation  \n",
       "2                  14988402;15509807;16858867;20627952                PMID  \n",
       "3                                             10082549  default activation  \n",
       "4    10330138;10602487;15531578;17629633;8663310;92...                PMID  \n",
       "..                                                 ...                 ...  \n",
       "362                                           19635798                PMID  \n",
       "363                         12438259;12665598;18347061                PMID  \n",
       "364                                           24041570                PMID  \n",
       "365                                           24041570                PMID  \n",
       "366                                           24041570                PMID  \n",
       "\n",
       "[367 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dc.op.collectri(organism=\"s.cerevisiae\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ckc]",
   "language": "python",
   "name": "conda-env-ckc-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
