{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a830bf8d",
   "metadata": {},
   "source": [
    "# Conversion to other organisms\n",
    "\n",
    "Most of the prior knowledge stored inside `Omnipath` is derived from human data, therefore they use gene names. Despite this, using homology we can convert gene names to other organisms.\n",
    "\n",
    "To showcase how to do it inside `decoupler`, we will load the `MSigDB` database and convert it into gene symbols for mouse and fly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b70696b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>collection</th>\n",
       "      <th>geneset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSC</td>\n",
       "      <td>oncogenic_signatures</td>\n",
       "      <td>PKCA_DN.V1_DN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSC</td>\n",
       "      <td>mirna_targets</td>\n",
       "      <td>MIR12123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSC</td>\n",
       "      <td>chemical_and_genetic_perturbations</td>\n",
       "      <td>NIKOLSKY_BREAST_CANCER_8Q12_Q22_AMPLICON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSC</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>GSE32986_UNSTIM_VS_GMCSF_AND_CURDLAN_LOWDOSE_S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSC</td>\n",
       "      <td>chemical_and_genetic_perturbations</td>\n",
       "      <td>BENPORATH_PRC2_TARGETS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407729</th>\n",
       "      <td>OR2W5P</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>GSE22601_DOUBLE_NEGATIVE_VS_CD8_SINGLE_POSITIV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407730</th>\n",
       "      <td>OR2W5P</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>KANNAN_BLOOD_2012_2013_TIV_AGE_65PLS_REVACCINA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407731</th>\n",
       "      <td>OR52L2P</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>GSE22342_CD11C_HIGH_VS_LOW_DECIDUAL_MACROPHAGE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407732</th>\n",
       "      <td>CSNK2A3</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>OCONNOR_PBMC_MENVEO_ACWYVAX_AGE_30_70YO_7DY_AF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407733</th>\n",
       "      <td>AQP12B</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>MATSUMIYA_PBMC_MODIFIED_VACCINIA_ANKARA_VACCIN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2407734 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genesymbol                          collection  \\\n",
       "0              MSC                oncogenic_signatures   \n",
       "1              MSC                       mirna_targets   \n",
       "2              MSC  chemical_and_genetic_perturbations   \n",
       "3              MSC              immunologic_signatures   \n",
       "4              MSC  chemical_and_genetic_perturbations   \n",
       "...            ...                                 ...   \n",
       "2407729     OR2W5P              immunologic_signatures   \n",
       "2407730     OR2W5P              immunologic_signatures   \n",
       "2407731    OR52L2P              immunologic_signatures   \n",
       "2407732    CSNK2A3              immunologic_signatures   \n",
       "2407733     AQP12B              immunologic_signatures   \n",
       "\n",
       "                                                   geneset  \n",
       "0                                            PKCA_DN.V1_DN  \n",
       "1                                                 MIR12123  \n",
       "2                 NIKOLSKY_BREAST_CANCER_8Q12_Q22_AMPLICON  \n",
       "3        GSE32986_UNSTIM_VS_GMCSF_AND_CURDLAN_LOWDOSE_S...  \n",
       "4                                   BENPORATH_PRC2_TARGETS  \n",
       "...                                                    ...  \n",
       "2407729  GSE22601_DOUBLE_NEGATIVE_VS_CD8_SINGLE_POSITIV...  \n",
       "2407730  KANNAN_BLOOD_2012_2013_TIV_AGE_65PLS_REVACCINA...  \n",
       "2407731  GSE22342_CD11C_HIGH_VS_LOW_DECIDUAL_MACROPHAGE...  \n",
       "2407732  OCONNOR_PBMC_MENVEO_ACWYVAX_AGE_30_70YO_7DY_AF...  \n",
       "2407733  MATSUMIYA_PBMC_MODIFIED_VACCINIA_ANKARA_VACCIN...  \n",
       "\n",
       "[2407734 rows x 3 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import decoupler as dc\n",
    "\n",
    "msigdb = dc.get_resource('MSigDB')\n",
    "msigdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50ab6a57",
   "metadata": {},
   "source": [
    "For this example we will filter by the `hallmark` gene sets collection:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad44cfa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>collection</th>\n",
       "      <th>geneset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MSC</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>ICOSLG</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>ICOSLG</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_INFLAMMATORY_RESPONSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>ICOSLG</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_ALLOGRAFT_REJECTION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>FOSL2</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_HYPOXIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878342</th>\n",
       "      <td>FOXO1</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878418</th>\n",
       "      <td>GCG</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878512</th>\n",
       "      <td>PDX1</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878605</th>\n",
       "      <td>INS</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878785</th>\n",
       "      <td>SRP9</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7318 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genesymbol collection                           geneset\n",
       "11            MSC   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "149        ICOSLG   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "223        ICOSLG   hallmark    HALLMARK_INFLAMMATORY_RESPONSE\n",
       "270        ICOSLG   hallmark      HALLMARK_ALLOGRAFT_REJECTION\n",
       "398         FOSL2   hallmark                  HALLMARK_HYPOXIA\n",
       "...           ...        ...                               ...\n",
       "878342      FOXO1   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "878418        GCG   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "878512       PDX1   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "878605        INS   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "878785       SRP9   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "\n",
       "[7318 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter by hallmark\n",
    "msigdb = msigdb[msigdb['collection']=='hallmark']\n",
    "\n",
    "# Remove duplicated entries\n",
    "msigdb = msigdb[~msigdb.duplicated(['geneset', 'genesymbol'])]\n",
    "msigdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f3ac365",
   "metadata": {},
   "source": [
    "Then, we can easily transform the obtained resource into mouse genes. Organisms can be defined by their common name, latin name or [NCBI Taxonomy identifier](https://www.ncbi.nlm.nih.gov/taxonomy).\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note**\n",
    "    \n",
    "Translating to an organism for the first time might take a while (~ 15 minutes). Since the data is stored in cache, the next times is going to run faster. If you need to reset the cache, run `rm -r .pypath/cache/`.\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "819128fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>collection</th>\n",
       "      <th>geneset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Msc</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Icoslg</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Icoslg</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_INFLAMMATORY_RESPONSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Icoslg</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_ALLOGRAFT_REJECTION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fosl2</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_HYPOXIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7710</th>\n",
       "      <td>Gcg</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7711</th>\n",
       "      <td>Pdx1</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7712</th>\n",
       "      <td>Ins2</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7713</th>\n",
       "      <td>Ins1</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7714</th>\n",
       "      <td>Srp9</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7644 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genesymbol collection                           geneset\n",
       "0           Msc   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "1        Icoslg   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "2        Icoslg   hallmark    HALLMARK_INFLAMMATORY_RESPONSE\n",
       "3        Icoslg   hallmark      HALLMARK_ALLOGRAFT_REJECTION\n",
       "4         Fosl2   hallmark                  HALLMARK_HYPOXIA\n",
       "...         ...        ...                               ...\n",
       "7710        Gcg   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "7711       Pdx1   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "7712       Ins2   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "7713       Ins1   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "7714       Srp9   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "\n",
       "[7644 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translate targets\n",
    "mouse_msigdb = dc.translate_net(msigdb, target_organism = 'mouse', unique_by = ('geneset', 'genesymbol'))\n",
    "mouse_msigdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e9cfadb",
   "metadata": {},
   "source": [
    "Note that when performing homology convertion we might gain or lose some genes from one organism to another.\n",
    "\n",
    "Let us try the fruit fly (`7227`) now:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9403ea9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>collection</th>\n",
       "      <th>geneset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hand</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CG12648</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>twi</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLH54F</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dl</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6594</th>\n",
       "      <td>Dmel\\CG6484</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6598</th>\n",
       "      <td>G6P</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6599</th>\n",
       "      <td>amon</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6600</th>\n",
       "      <td>Cbp53E</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6601</th>\n",
       "      <td>Srp9</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5868 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genesymbol collection                           geneset\n",
       "0            Hand   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "1         CG12648   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "2             twi   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "3          HLH54F   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "4              dl   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "...           ...        ...                               ...\n",
       "6594  Dmel\\CG6484   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "6598          G6P   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "6599         amon   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "6600       Cbp53E   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "6601         Srp9   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "\n",
       "[5868 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translate targets\n",
    "fly_msigdb = dc.translate_net(msigdb, target_organism = 7227, unique_by = ('genesymbol', 'geneset'))\n",
    "fly_msigdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbf238d3",
   "metadata": {},
   "source": [
    "The `translate_net` function provides finer control, but in most cases it's enough to pass the name of the desired organism to the functions that download the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1ee435a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cd247</td>\n",
       "      <td>T-cell receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tab2</td>\n",
       "      <td>JAK/STAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tab2</td>\n",
       "      <td>Receptor tyrosine kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tab2</td>\n",
       "      <td>Toll-like receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tab2</td>\n",
       "      <td>Innate immune pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1017</th>\n",
       "      <td>Sit1</td>\n",
       "      <td>T-cell receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1018</th>\n",
       "      <td>Nfatc2</td>\n",
       "      <td>T-cell receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1019</th>\n",
       "      <td>Nfatc2</td>\n",
       "      <td>B-cell receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1020</th>\n",
       "      <td>Rasgrp1</td>\n",
       "      <td>T-cell receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1021</th>\n",
       "      <td>Rasgrp1</td>\n",
       "      <td>B-cell receptor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1022 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genesymbol                   pathway\n",
       "0         Cd247           T-cell receptor\n",
       "1          Tab2                  JAK/STAT\n",
       "2          Tab2  Receptor tyrosine kinase\n",
       "3          Tab2        Toll-like receptor\n",
       "4          Tab2    Innate immune pathways\n",
       "...         ...                       ...\n",
       "1017       Sit1           T-cell receptor\n",
       "1018     Nfatc2           T-cell receptor\n",
       "1019     Nfatc2           B-cell receptor\n",
       "1020    Rasgrp1           T-cell receptor\n",
       "1021    Rasgrp1           B-cell receptor\n",
       "\n",
       "[1022 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spw = dc.get_resource('SignaLink_pathway', organism = 'rat')\n",
    "spw"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7eff89d",
   "metadata": {},
   "source": [
    "PROGENy and DoRothEA have their own dedicated functions which work a similar way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6239f93d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Androgen</td>\n",
       "      <td>Tmprss2</td>\n",
       "      <td>11.490631</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Androgen</td>\n",
       "      <td>Nkx3-1</td>\n",
       "      <td>10.622551</td>\n",
       "      <td>2.242078e-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NFkB</td>\n",
       "      <td>Nkx3-1</td>\n",
       "      <td>2.372983</td>\n",
       "      <td>5.589476e-32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TNFa</td>\n",
       "      <td>Nkx3-1</td>\n",
       "      <td>2.871633</td>\n",
       "      <td>1.044050e-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Androgen</td>\n",
       "      <td>Mboat2</td>\n",
       "      <td>10.472733</td>\n",
       "      <td>4.624285e-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1475</th>\n",
       "      <td>p53</td>\n",
       "      <td>Carns1</td>\n",
       "      <td>4.538734</td>\n",
       "      <td>4.730570e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>p53</td>\n",
       "      <td>Ccdc150</td>\n",
       "      <td>-3.174527</td>\n",
       "      <td>7.396252e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1477</th>\n",
       "      <td>p53</td>\n",
       "      <td>Trem2</td>\n",
       "      <td>4.101937</td>\n",
       "      <td>9.739648e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1478</th>\n",
       "      <td>p53</td>\n",
       "      <td>Gdf9</td>\n",
       "      <td>3.355741</td>\n",
       "      <td>1.087433e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1479</th>\n",
       "      <td>p53</td>\n",
       "      <td>Nhlh2</td>\n",
       "      <td>2.201638</td>\n",
       "      <td>1.651582e-12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1447 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        source   target     weight       p_value\n",
       "0     Androgen  Tmprss2  11.490631  0.000000e+00\n",
       "1     Androgen   Nkx3-1  10.622551  2.242078e-44\n",
       "2         NFkB   Nkx3-1   2.372983  5.589476e-32\n",
       "3         TNFa   Nkx3-1   2.871633  1.044050e-27\n",
       "4     Androgen   Mboat2  10.472733  4.624285e-44\n",
       "...        ...      ...        ...           ...\n",
       "1475       p53   Carns1   4.538734  4.730570e-13\n",
       "1476       p53  Ccdc150  -3.174527  7.396252e-13\n",
       "1477       p53    Trem2   4.101937  9.739648e-13\n",
       "1478       p53     Gdf9   3.355741  1.087433e-12\n",
       "1479       p53    Nhlh2   2.201638  1.651582e-12\n",
       "\n",
       "[1447 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dc.get_progeny(organism = 'Mus musculus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "82fec8b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>confidence</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E2f4</td>\n",
       "      <td>A</td>\n",
       "      <td>Mycl</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tp53</td>\n",
       "      <td>A</td>\n",
       "      <td>Ogg1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E2f4</td>\n",
       "      <td>A</td>\n",
       "      <td>Bach1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hif1a</td>\n",
       "      <td>A</td>\n",
       "      <td>Mif</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E2f4</td>\n",
       "      <td>A</td>\n",
       "      <td>Aurkb</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30885</th>\n",
       "      <td>Irf4</td>\n",
       "      <td>C</td>\n",
       "      <td>Jak1</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30886</th>\n",
       "      <td>Irf4</td>\n",
       "      <td>C</td>\n",
       "      <td>Il16</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30887</th>\n",
       "      <td>Irf4</td>\n",
       "      <td>C</td>\n",
       "      <td>Ikzf3</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30888</th>\n",
       "      <td>Irf4</td>\n",
       "      <td>C</td>\n",
       "      <td>Lpp</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30889</th>\n",
       "      <td>Znf740</td>\n",
       "      <td>C</td>\n",
       "      <td>Znf687</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30890 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       source confidence  target    weight\n",
       "0        E2f4          A    Mycl  1.000000\n",
       "1        Tp53          A    Ogg1  1.000000\n",
       "2        E2f4          A   Bach1  1.000000\n",
       "3       Hif1a          A     Mif  1.000000\n",
       "4        E2f4          A   Aurkb  1.000000\n",
       "...       ...        ...     ...       ...\n",
       "30885    Irf4          C    Jak1  0.333333\n",
       "30886    Irf4          C    Il16  0.333333\n",
       "30887    Irf4          C   Ikzf3  0.333333\n",
       "30888    Irf4          C     Lpp  0.333333\n",
       "30889  Znf740          C  Znf687  0.333333\n",
       "\n",
       "[30890 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dc.get_dorothea(organism = 'mouse')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
